pandas/tests/dtypes/cast/test_promote.py

```diff
@@ -860,3 +1011,29 @@ def test_maybe_promote_dimensions(any_numpy_dtype_reduced, dim):
     assert (result_missing_value == expected_missing_value) or (
         result_missing_value is np.nan and expected_missing_value is np.nan
     )
+
+    # same again for _maybe_promote_with_array (for coverage)
+    result_dtype, result_missing_value = _maybe_promote_with_array(dtype, fill_array)
+
+    assert result_dtype == expected_dtype
+    # None == None, iNaT == iNaT, but np.nan != np.nan
+    assert (result_missing_value == expected_missing_value) or (
+        result_missing_value is np.nan and expected_missing_value is np.nan
```

Comment:
is this specifically np.nan or is it a more general "same numpy NA value", e.g. matching `np.datetime64("nat")`?

Reply:
This is only `np.nan`, because no NA-values other than (`np.nan`, `pd.Nat`, `iNaT` and `None`) are tested in this module

Reply:
OK.  Note that np.nan is not a singleton the way pd.NaT is, so e.g. `np.float64(np.nan)` is not the same object as `np.nan`.  Is it possible to get any _other_ NaN?

Reply:
It would be easy to add more, yes. It's just parametrization in all the tests ending in `_with_na`. Since the actual promotion code uses `isna`, I don't expect any problems from this (fingers crossed).

Do you have a list of NA-values you'd want me to try?

Reply:
I'm still working through these tests, so you probably know the desired behavior better than I do.  Best guess is that this should use the same pattern as in #28561 with checking `type(x) == type(y)` and then checking `isna(x) and isna(y)`